packages/puppeteer-core/src/common/Input.ts

```diff
@@ -438,90 +513,178 @@ export class Mouse {
+   * Presses the mouse.
+   *
+   * @param options - Options to configure behavior.
    */
   async down(options: MouseOptions = {}): Promise<void> {
-    const {button = 'left', clickCount = 1} = options;
-    this.#button = button;
-    await this.#client.send('Input.dispatchMouseEvent', {
-      type: 'mousePressed',
-      button,
-      x: this.#x,
-      y: this.#y,
-      modifiers: this.#keyboard._modifiers,
-      clickCount,
+    const {button = MouseButton.Left, clickCount = 1} = options;
+    const flag = getFlag(button);
+    if (!flag) {
+      throw new Error(`Unsupported mouse button: ${button}`);
+    }
+    if (this.#state.buttons & flag) {
+      throw new Error(`'${button}' is already pressed.`);
+    }
+    await this.#withTransaction(updateState => {
+      updateState({buttons: this.#state.buttons | flag});
+      const {buttons, position} = this.#state;
+      return this.#client.send('Input.dispatchMouseEvent', {
+        type: 'mousePressed',
+        modifiers: this.#keyboard._modifiers,
+        clickCount,
+        buttons,
+        button,
+        ...position,
+      });
     });
   }
 
   /**
-   * Dispatches a `mouseup` event.
-   * @param options - Optional `MouseOptions`.
+   * Releases the mouse.
+   *
+   * @param options - Options to configure behavior.
    */
   async up(options: MouseOptions = {}): Promise<void> {
-    const {button = 'left', clickCount = 1} = options;
-    this.#button = 'none';
-    await this.#client.send('Input.dispatchMouseEvent', {
-      type: 'mouseReleased',
-      button,
-      x: this.#x,
-      y: this.#y,
-      modifiers: this.#keyboard._modifiers,
-      clickCount,
+    const {button = MouseButton.Left, clickCount = 1} = options;
+    const flag = getFlag(button);
+    if (!flag) {
+      throw new Error(`Unsupported mouse button: ${button}`);
+    }
+    if (!(this.#state.buttons & flag)) {
+      throw new Error(`'${button}' is not pressed.`);
+    }
+    await this.#withTransaction(updateState => {
+      updateState({buttons: this.#state.buttons & ~flag});
+      const {buttons, position} = this.#state;
+      return this.#client.send('Input.dispatchMouseEvent', {
+        type: 'mouseReleased',
+        modifiers: this.#keyboard._modifiers,
+        clickCount,
+        buttons,
+        button,
+        ...position,
+      });
     });
   }
 
+  /**
+   * Shortcut for `mouse.move`, `mouse.down` and `mouse.up`.
+   *
+   * @param x - Horizontal position of the mouse.
+   * @param y - Vertical position of the mouse.
+   * @param options - Options to configure behavior.
+   */
+  async click(
+    x: number,
+    y: number,
+    options: MouseClickOptions = {}
+  ): Promise<void> {
+    const {delay} = options;
+    const actions: Array<Promise<void>> = [];
+    const {position} = this.#state;
+    if (position.x !== x || position.y !== y) {
+      actions.push(this.move(x, y));
+    }
+    actions.push(this.down(options));
+    if (typeof delay === 'number') {
```

Comment:
nit: it could be simplified I think so that actions don't need to be cleared:

```
const delayPromise = typeof delay === 'number' ? ...timeout : Promise.resolve();
actions.push(delayPromise.then(() => this.up(options)));
```

Reply:
Ack